Improvements for CarCRUD

manu3210 opened the pull request:
## Background

I wanted to improve the code adding validations and erasing unnecessary lines. 
I would like you can tell me if this code is right or what could I change to improve it. In the future, I want to convert this project in a MVC web application.

## Changes

- Added a validation for Create method, verifying if the car that is passed as parameter is null.
- Many empty lines was deleted.

manu3210 committed:
Improvements for CarCRUD

jrocaspana commented on CarRentals/CarRentals/Program.cs:
```diff
@@ -16,14 +16,12 @@ static void Main(string[] args)
             
             CarCRUD option = new CarCRUD();
```
maybe we can find a better name for this object? also, remember to use implicit types when possible.

manu3210 replied:
Changed it

jrocaspana commented on CarRentals/CarRentals/Controllers/CarCRUD.cs:
```diff
@@ -87,7 +81,6 @@ public void Delete(int id)
         }
 
-
         public void ReadJson()
```
any reason to make this method public?
also, I'm not sure about having this try catch here, what do you think the desired behavior should be if there's an error while opening/reading the JSON file? and how's that compared to the current behavior?

manu3210 replied:
I corrected the acces modifier to private in this method and in SaveChanges()

jrocaspana commented on CarRentals/CarRentals/Controllers/CarCRUD.cs:
```diff
@@ -18,24 +18,20 @@ public class CarCRUD
         }
 
-
         public Car Get(int id)
```
throwing exceptions is expensive, we should be really picky about when and why. not finding a car may not be enough to throw one, perhaps we can just return null? what do you think?

manu3210 replied:
Thanks! I have changed all this lines in only one, deleting exceptions and putting a LINQ lambda expression.

jrocaspana commented on CarRentals/CarRentals/Controllers/CarCRUD.cs:
```diff
@@ -57,7 +53,6 @@ public Car Get(int id)
         }
 
-
         public Car Update(Car car)
```
I see here you implemented a `for` statement and in the Get you used a `foreach`, in both you never stop iterating after finding the element meaning if the element you are looking for is the first of a sequence of 1000 elements, you would still iterate through all of them before returning the result. 
also, a suggestion: feel free to investigate about LINQ (https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/introduction-to-linq-queries) and see if you can implement it here

jrocaspana reviewed (commented):
looks really good! nice work!
I left some comments. Also, try to implement and take into account what we shared today in the classroom (https://classroom.google.com/c/Mzc4NTA4NzUzOTkw/p/Mzg3Njg3OTk0Mzc5/details)

manu3210 committed:
Deleted exceptions, added "usings" and app.Config

Also:

- I have used a LINQ lambda expression instead of foreach in the Get method

-  Corrected some details in the code like implicit variables and access modifiers

jrocaspana commented on CarRentals/CarRentals/Controllers/CarCRUD.cs:
```diff
@@ -1,80 +1,60 @@
-                {
-                    CarList[i] = car;
-                }
+                CarList[CarList.IndexOf(getCar)] = car;
```
can you think in any other way to update the car that doesn't involve using the element index?

manu3210 replied:
Thanks for your reply. Sincerily I've thought in other ways to do that but I think that it's worst.
I thought calling to the methods delete and after create to replace the item without use the index, but I believe that is worst than use it. 
I tried with LINQ, but I didn't find any statement that could help me.
However I'm going to continue reviewing this method to improve it.

jrocaspana replied:
that's great! did you try just fetching the object from the list and updating it? not assigning one object to the other like `obj1 = car` (same thing you are doing now), but updating the properties? would look like something like this (pseudo code):
```csharp
var existingCar = CarList.Where(..).FirstOrDefault();
existingCar.Doors = car.Doors;
// do the same of other properties
// profit
```

we can discuss today in our meeting if this worked or not

jrocaspana reviewed (commented):
this is getting better and better! keep up the great work!

jrocaspana commented on CarRentals/CarRentals/Controllers/CarCRUD.cs:
```diff
@@ -1,80 +1,60 @@
+    public class CarCRUD : IDataProcessing
     {
-        const string PATH = @"Cars.json";
+        string PATH = ConfigurationManager.AppSettings["JsonFile"].ToString();
```
this can be improved with https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-5.0

manu3210 committed:
Method Update was corrected

jrocaspana commented on CarRentals/CarRentals/Controllers/CarCRUD.cs:
```diff
@@ -13,7 +13,7 @@ namespace CarRentals
     public class CarCRUD : IDataProcessing
     {
-        string PATH = ConfigurationManager.AppSettings["JsonFile"].ToString();
+        string Path = Program.JsonFilePath();
```
ultra minor: unless this is a public property, by code convention it should start with lowercase and with a `_` as a prefix (check https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#camel-case)

manu3210 replied:
Thanks! I have corrected it. However I'm not happy with this, because I think that you want I use dependency injection. I tried to watch a lot of videos and read a lot of post and the Microsoft documentation, but I couldn't achieve this in a console app. I'm not going to give up, I will continue trying. Anyway, when I will change the console for a web application, it will be easyest

manu3210 replied:
@jrocaspana I have created a new pull request with a little change in the project. And now I think that is much better. 
This is the link: 
https://github.com/manu3210/MSAcademy/pull/2

Please, could you tell me what you think?.

jrocaspana replied:
sure! I will check today later in the morning and give you feedback, we can also discuss this topic today in our meeting

jrocaspana reviewed (approved):
beautiful! well done Manu!

manu3210 committed:
_path instead of Path
